### PR TITLE
Update BasicAuthenticationHandler.cs

### PR DESCRIPTION
**Pull request by josea**
Hi,

I was having an issue with the Web Api and AllowAnonymous. It is explained in detail [here](http://stackoverflow.com/questions/15953291/webapidoodle-basicauthenticationhandler-and-webapi-4-not-allowing-anonymous/15956342?noredirect=1#15956342).

The change it to invoke the Web Api. I believe it is safe.

### CODE REVIEW

**Comment by tugberkugurlu** [comment]
Hi josea,

This's by design. `HandleUnauthenticatedRequest` is a `virtual` method and you should override it to change the default behavior.

Also, you have a bug in your code. I'll comment on your file.

File: src/WebApiDoodle.Web/MessageHandlers/BasicAuthenticationHandler.cs [hunk 1]
```diff
@@ -68,7 +68,21 @@ public abstract class BasicAuthenticationHandler : DelegatingHandler {
...
 
                 // Request is not authanticated. Handle unauthenticated request.
-                return HandleUnauthenticatedRequestImpl(request, cancellationToken);
+                //return HandleUnauthenticatedRequestImpl(request, cancellationToken);
+                // fixed to work with WebApi 
+                // http://stackoverflow.com/questions/15953291/webapidoodle-basicauthenticationhandler-and-webapi-4-not-allowing-anonymous/15956342?noredirect=1#15956342
+                return base.SendAsync(request, cancellationToken).ContinueWith<HttpResponseMessage>(
+                    task =>
+                    {
+                        var response = task.Result ;
```
Comment (tugberkugurlu) [inline src/WebApiDoodle.Web/MessageHandlers/BasicAuthenticationHandler.cs:77]: This line will potentially throw an exception depending on the status of the task.

**Comment by josea** [comment]
I see. I'll play around with it. Note: wouldn't it be nice for the library to work with WebApi by default?. It is kind of confusing that the AllowAnonymous attribute just stops working in the WebApi ApiController (I can't imagine who would like that behavior). Anyways, I appreciate your observations and will implement accordingly. Thanks.

BTW: I'll also update the SO entry in case somebody else gets into this situation.

**Comment by tugberkugurlu** [comment]
@josea 

>  Note: wouldn't it be nice for the library to work with WebApi by default?. It is kind of confusing that the AllowAnonymous attribute just stops working in the WebApi ApiController.

`AllowAnonymousAttribute` has no logic inside it. `AuthorizeAttribute` honors it. In my personal opinion, every user should authenticate to an API if that API needs authentication. That's why u see this as the default behavior. However, the fix is easy. Just override `HandleUnauthenticatedRequest` method and keep it empty. Then, add the challenge request as a continuation to your handler's SendAsync method just like you did.

**Comment by josea** [comment]
Fair enough. Thanks - sorry to have taken your time (I actually meant to help).

**Comment by tugberkugurlu** [comment]
A healthy discussion is always good :) 

Thanks for the pull request and comments.

**Pull request closed without merging**